nexus/src/app/background/plan_execution.rs

```diff
@@ -0,0 +1,327 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for realizing a plan blueprint
+
+use super::common::BackgroundTask;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::stream;
+use futures::FutureExt;
+use futures::StreamExt;
+use nexus_db_queries::context::OpContext;
+use nexus_types::blueprint::{Blueprint, OmicronZonesConfig};
+use serde_json::json;
+use slog::Logger;
+use std::collections::BTreeMap;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+/// Background task that takes a [`Blueprint`] and realizes the change to
+/// the state of the system based on the `Blueprint`.
+pub struct PlanExecutor {
+    rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+}
+
+impl PlanExecutor {
+    // Temporary until we wire up the background task
+    #[allow(unused)]
+    pub fn new(
+        rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+    ) -> PlanExecutor {
+        PlanExecutor { rx_blueprint }
+    }
+}
+
+impl BackgroundTask for PlanExecutor {
+    fn activate<'a, 'b, 'c>(
+        &'a mut self,
+        opctx: &'b OpContext,
+    ) -> BoxFuture<'c, serde_json::Value>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        async {
+            // Get the latest blueprint, cloning to prevent holding a read lock
+            // on the watch.
+            let blueprint = self.rx_blueprint.borrow().clone();
+
+            let Some(blueprint) = blueprint else {
+                warn!(&opctx.log,
+                      "Plan execution: skipped";
+                      "reason" => "no blueprint");
+                return json!({"error": "no blueprint" });
```

Comment:
Is this the right design for the trait method? I thought maybe this would return a `Result` type (though maybe this was discussed already).

Reply:
The caller doesn't do anything different with the outcome.  (If that's surprising, the [design notes](https://github.com/oxidecomputer/omicron/blob/255cb83822098f80c2f60811152922259ca8e82c/nexus/src/app/background/common.rs#L65-L129) may shed some light on that part.)  And often it's not obvious how to classify the outcome.  Using DNS propagation as the canonical example: the task may propagate the latest DNS config to 3 out of 5 servers.  That's not obviously either a success or failure -- and fortunately, we don't need to decide since the caller doesn't care.  The most important consumer of this is a human looking at the system to understand what it did: "DNS generation X was propagated to these servers, but we couldn't reach these other servers and here are the errors we encountered talking to each one."  [For that kind of summary, the success case looks pretty similar to the case where some errors were encountered.](https://github.com/oxidecomputer/omicron/blob/255cb83822098f80c2f60811152922259ca8e82c/nexus/src/app/background/dns_propagation.rs#L93-L125)  So right now this is just a JSON value to give maximum flexibility.  (omdb just knows how to interpret those.)  I don't love it -- obviously this could be a more typed value, but the cost seemed to me to outweigh the benefit here.  If we end up wanting programmatic consumers of this information, we might want to revisit, but then we'll have to deal with the fact that background tasks return different types, even though we want to treat them uniformly, etc.

(I think this is also outside the scope of the PR, but it's a good question!)

Reply:
That makes sense! Sorry for the distraction; this is definitely a hard problem.